### PR TITLE
otelgrpc: Fix NewServerHandler to correctly set span status

### LINKED ISSUE #4586: otelgrpc: server stats handler sets all errors to error=true
### Description

Any server gRPC error codes are treated with `error=true`,  according to [gRPC status spec](https://opentelemetry.io/docs/specs/semconv/rpc/grpc/#grpc-status) only certain server gRPC status codes should set `error=true`.

In the interceptor we check the serverStatus with https://github.com/open-telemetry/opentelemetry-go-contrib/blob/ba19074a6785b2b65c86a52c0649cf040319fa96/instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go#L328-L330 (correct behavior)

stats handler does not call `serverStatus` for server spans to check if error should be set to true: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/ba19074a6785b2b65c86a52c0649cf040319fa96/instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go#L187-L189 (incorrect behavior)

(Need to check if it is server or client stats handler to determine if we want to call `serverStatus`)

### Environment

- OS: macOS
- Architecture: [e.g. x86, i386]
- Go Version: 1.21
- `otelgrpc` version: v0.46.1

### Steps To Reproduce

1. Write a grpc endpoint that returns `PERMISSION_DENIED`
2. Use `grpc.NewServer(grpc.StatsHandler(otelgrpc.NewServerHandler()))`
3. Call grpc endpoint
4. Trace has error set to true
resulting to this (from stdout)
```
"Status": {
	"Code": "Error",
	"Description": "PERMISSION_DENIED"
},
```

### Expected behavior

If I switch to using `grpc.NewServer(grpc.UnaryInterceptor(otelgrpc.UnaryServerInterceptor()))`
I get the expected behavior:
```
"Status": {
	"Code": "Unset",
	"Description": ""
},
```

### PR DESCRIPTION
**Pull request by scorpionknifes**
fix #4586

### CODE REVIEW

**Review by pellared — CHANGES_REQUESTED** [review]
The changes are not compliant with https://opentelemetry.io/docs/specs/semconv/rpc/grpc/#grpc-status

File: instrumentation/google.golang.org/grpc/otelgrpc/config.go [hunk 1]
```diff
@@ -113,7 +113,7 @@ func newConfig(opts []Option, role string) *config {
 	}
 
 	c.rpcResponsesPerRPC, err = c.meter.Int64Histogram("rpc."+role+".responses_per_rpc",
-		metric.WithDescription("Measures the number of messages received per RPC. Should be 1 for all non-streaming RPCs."),
+		metric.WithDescription("Measures the number of messages responded per RPC. Should be 1 for all non-streaming RPCs."),
```
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/config.go:116]: The description. Was correct. See: https://opentelemetry.io/docs/specs/semconv/rpc/rpc-metrics/

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go [hunk 2]
```diff
@@ -85,7 +85,7 @@ func (h *serverHandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) cont
 
 // HandleRPC processes the RPC stats.
 func (h *serverHandler) HandleRPC(ctx context.Context, rs stats.RPCStats) {
-	h.handleRPC(ctx, rs)
+	h.handleRPC(ctx, rs, true)
```
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:88]: I do not find it readable.

I propose the following

```suggestion
	isServer := true
	h.handleRPC(ctx, rs, isServer)
```

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go [hunk 3]
```diff
@@ -121,7 +121,7 @@ func (h *clientHandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) cont
 
 // HandleRPC processes the RPC stats.
 func (h *clientHandler) HandleRPC(ctx context.Context, rs stats.RPCStats) {
-	h.handleRPC(ctx, rs)
+	h.handleRPC(ctx, rs, false)
```
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:124]: I do not find it readable.

I propose the following.

```suggestion
	isServer := false
	h.handleRPC(ctx, rs, isServer)
```

**Review by pellared — COMMENTED** [review]
Overall looks good 👍 Few tiny comments

File: CHANGELOG.md [hunk 4]
```diff
@@ -12,6 +12,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - Add the new `go.opentelemetry.io/contrib/instrgen` package to provide auto-generated source code instrumentation. (#3068, #3108)
 
+### Fixed
+
+- Fix `NewServerHandler` in `go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc` sets error state from grpc status incorrectly (#4587)
```
Comment (pellared) [inline CHANGELOG.md:17]:
```suggestion
- Fix `NewServerHandler` in `go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc` to correctly set the span as error depending on the gRPC status. (#4587)
```

Comment (pellared) [inline CHANGELOG.md:17]:
```suggestion
- Fix `NewServerHandler` in `go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc` to correctly set the span as error depending on the gRPC status. (#4587)
```

File: instrumentation/google.golang.org/grpc/otelgrpc/test/stats_handler_test.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata/metricdatatest"
+	"go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/sdk/trace/tracetest"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
+	grpc_codes "google.golang.org/grpc/codes"
+	"google.golang.org/grpc/stats"
+	"google.golang.org/grpc/status"
```
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/test/stats_handler_test.go:33]: This is not "gofumpted" (see build error).

**Review by pellared — APPROVED** [review]
Thank you 🥇

**Review by dmathieu — APPROVED** [review]

**Review by dashpole — APPROVED** [review]

**Pull request merged**